File: app/src/main/java/com/plweegie/android/squash/RepoListFragment.java
/*
 * Copyright (c) 2017 Jan K Szymanski

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
package com.plweegie.android.squash;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.plweegie.android.squash.adapters.RepoAdapter;
import com.plweegie.android.squash.data.RepoEntry;
import com.plweegie.android.squash.data.RepoRepository;
import com.plweegie.android.squash.rest.GitHubService;
import com.plweegie.android.squash.utils.Injectors;
import com.plweegie.android.squash.utils.PaginationScrollListener;
import com.plweegie.android.squash.utils.QueryPreferences;

import java.util.List;

import javax.inject.Inject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class RepoListFragment extends Fragment implements RepoAdapter.RepoAdapterOnClickHandler {

    private static final int MAXIMUM_LIST_LENGTH = 30;
    private static final int START_PAGE = 1;

    @Inject
    GitHubService mService;

    @Inject
    SharedPreferences mSharedPrefs;

    @Inject
    QueryPreferences mQueryPrefs;

    private boolean isLoading = false;
    private boolean isLastPage = false;
    private int currentPage = START_PAGE;

    @Inject
    RepoRepository mDataRepository;
    private RecyclerView mRecyclerView;
    private LinearLayoutManager mManager;
    private RepoAdapter mAdapter;
    private ProgressBar mIndicator;
    private InputMethodManager mImm;

    private SharedPreferences.OnSharedPreferenceChangeListener mListener;

    public static RepoListFragment newInstance() {
        return new RepoListFragment();
    }
    
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        mListener = new SharedPreferences.OnSharedPreferenceChangeListener() {
            @Override
            public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String s) {
                mAdapter.sort();
            }
        };
    }

    @Override
    public void onAttach(Context context) {
        ((App) getActivity().getApplication()).getNetComponent().inject(this);
        super.onAttach(context);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup parent,
            Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.list_fragment, parent, false);

        mImm = (InputMethodManager) getActivity().getSystemService(Context
                .INPUT_METHOD_SERVICE);
        mRecyclerView = v.findViewById(R.id.commits_recycler_view);
        mIndicator = v.findViewById(R.id.load_indicator);
        mIndicator.setVisibility(View.GONE);

        mAdapter = new RepoAdapter(getActivity(), this);
        mManager = new LinearLayoutManager(getActivity());
        
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(mManager);
        mRecyclerView.addItemDecoration(new DividerItemDecoration(getActivity(),
                LinearLayoutManager.VERTICAL));

        mRecyclerView.setAdapter(mAdapter);
        mRecyclerView.addOnScrollListener(new PaginationScrollListener(mManager) {
            @Override
            protected void loadMoreItems() {
                isLoading = true;
                currentPage++;
                updateUI();
            }

            @Override
            public int getTotalPageCount() {
                return MAXIMUM_LIST_LENGTH;
            }

            @Override
            public boolean isLastPage() {
                return isLastPage;
            }

            @Override
            public boolean isLoading() {
                return isLoading;
            }
        });

        updateUI();
        
        return v;
    }

    @Override
    public void onResume() {
        super.onResume();
        mSharedPrefs.registerOnSharedPreferenceChangeListener(mListener);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mSharedPrefs.unregisterOnSharedPreferenceChangeListener(mListener);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.repo_list_menu, menu);
        
        final MenuItem searchItem = menu.findItem(R.id.repo_search);
        final SearchView searchView = (SearchView) searchItem.getActionView();
        
        searchView.setQueryHint(getResources().getString(R.string.search_hint));
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String string) {
                mImm.hideSoftInputFromWindow(searchView.getWindowToken(), 0);
                searchView.setQuery("", false);
                searchView.setIconified(true);
                
                mRecyclerView.setVisibility(View.INVISIBLE);
                mIndicator.setVisibility(View.VISIBLE);
                
                mQueryPrefs.setStoredQuery(string);
                isLoading = false;
                isLastPage = false;
                currentPage = START_PAGE;

                mAdapter.clear();
                updateUI();

                return true;
            }

            @Override
            public boolean onQueryTextChange(String string) {
                return true;
            }
        });
        
        searchView.setOnSearchClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String query = mQueryPrefs.getStoredQuery();
                searchView.setQuery(query, false);
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.sort_by:
                Intent intent = new Intent(getActivity(), SettingsActivity.class);
                startActivity(intent);
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onItemClick(int position) {
        mDataRepository.addFavorite(mAdapter.getItem(position));
        Log.d("RepoListFragment", "pos " + position + " clicked");
    }
    
    private void updateUI() {

        final String apiQuery = mQueryPrefs.getStoredQuery();
        final String authToken = mQueryPrefs.getStoredAccessToken();
        Call<List<RepoEntry>> call = mService.getRepos(apiQuery, currentPage, authToken);
        call.enqueue(new Callback<List<RepoEntry>>() {

            @Override
            public void onResponse(Call<List<RepoEntry>> call,
                                   Response<List<RepoEntry>> response) {

                isLoading = false;

                if (response.body() == null) {
                    Toast.makeText(getActivity(), "No repositories found for " + apiQuery,
                            Toast.LENGTH_SHORT).show();
                    mIndicator.setVisibility(View.GONE);
                    return;
                }

                List<RepoEntry> repos = response.body();

                if (repos.size() > 0) {
                    mAdapter.addAll(repos);
                    mAdapter.sort();
                }

                if (repos.size() < MAXIMUM_LIST_LENGTH) {
                    isLastPage = true;
                }

                mRecyclerView.setVisibility(View.VISIBLE);
                mIndicator.setVisibility(View.GONE);
            }

            @Override
            public void onFailure(Call<List<RepoEntry>> call, Throwable t) {
                Log.e("RepoListFragment", "Retrofit error: " + t);
            }
        });
    }

}

File: app/src/main/java/com/plweegie/android/squash/ui/RepoListFragment.kt
/*
 * Copyright (c) 2017 Jan K Szymanski

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
package com.plweegie.android.squash.ui

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.*
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.crashlytics.android.Crashlytics
import com.plweegie.android.squash.App
import com.plweegie.android.squash.R
import com.plweegie.android.squash.adapters.BaseGithubAdapter
import com.plweegie.android.squash.adapters.RepoAdapter
import com.plweegie.android.squash.data.RepoEntry
import com.plweegie.android.squash.rest.GitHubService
import com.plweegie.android.squash.utils.PaginationScrollListener
import com.plweegie.android.squash.utils.QueryPreferences
import com.plweegie.android.squash.viewmodels.RepoListViewModel
import com.plweegie.android.squash.viewmodels.RepoListViewModelFactory
import kotlinx.android.synthetic.main.list_fragment.*
import javax.inject.Inject


class RepoListFragment : Fragment(), RepoAdapter.RepoAdapterOnClickHandler,
        BaseGithubAdapter.GithubAdapterOnClickListener {

    @Inject
    lateinit var service: GitHubService

    @Inject
    lateinit var sharedPrefs: SharedPreferences

    @Inject
    lateinit var queryPrefs: QueryPreferences

    @Inject
    lateinit var viewModelFactory: RepoListViewModelFactory

    private lateinit var viewModel: RepoListViewModel
    private lateinit var repoAdapter: RepoAdapter
    private lateinit var manager: LinearLayoutManager
    private lateinit var imm: InputMethodManager

    private var isContentLoading = false
    private var isContentLastPage = false
    private var currentPage = START_PAGE

    private var apiQuery: String? = null

    private val listener = SharedPreferences.OnSharedPreferenceChangeListener {
        _, _ -> repoAdapter.sort()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        viewModel = ViewModelProvider(this, viewModelFactory).get(RepoListViewModel::class.java)
    }

    override fun onAttach(context: Context) {
        (activity?.application as App).netComponent.inject(this)
        super.onAttach(context)
    }

    override fun onCreateView(inflater: LayoutInflater, parent: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.list_fragment, parent, false)

        imm = activity?.getSystemService(Context
                .INPUT_METHOD_SERVICE) as InputMethodManager

        repoAdapter = RepoAdapter(activity).apply {
            setOnAddFavoriteListener(this@RepoListFragment)
            setListener(this@RepoListFragment)
        }
        manager = LinearLayoutManager(activity)

        apiQuery = queryPrefs.storedQuery ?: ""
        fetchRepos(apiQuery!!, currentPage)

        return v
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        load_indicator?.visibility = View.GONE

        commits_recycler_view?.apply {
            setHasFixedSize(true)
            layoutManager = manager
            addItemDecoration(DividerItemDecoration(activity, LinearLayoutManager.VERTICAL))
            adapter = repoAdapter
            addOnScrollListener(object : PaginationScrollListener(manager) {
                override fun loadMoreItems() {
                    isContentLoading = true
                    currentPage++
                    Toast.makeText(activity, getString(R.string.loading_more), Toast.LENGTH_SHORT)
                            .show()
                    fetchRepos(apiQuery!!, currentPage)
                }

                override fun getTotalPageCount(): Int = MAXIMUM_LIST_LENGTH

                override fun isLastPage(): Boolean = isContentLastPage

                override fun isLoading(): Boolean = isContentLoading
            })
        }

        observeUI()
    }

    override fun onResume() {
        super.onResume()
        sharedPrefs.registerOnSharedPreferenceChangeListener(listener)
    }

    override fun onDestroy() {
        super.onDestroy()
        sharedPrefs.unregisterOnSharedPreferenceChangeListener(listener)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.repo_list_menu, menu)

        val searchItem = menu.findItem(R.id.repo_search)
        val searchView = searchItem.actionView as SearchView

        searchView.queryHint = resources.getString(R.string.search_hint)
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(string: String): Boolean {
                imm.hideSoftInputFromWindow(searchView.windowToken, 0)
                searchView.apply {
                    setQuery("", false)
                    isIconified = true
                }

                commits_recycler_view?.visibility = View.INVISIBLE
                load_indicator?.visibility = View.VISIBLE

                queryPrefs.storedQuery = string
                apiQuery = queryPrefs.storedQuery

                isContentLoading = false
                isContentLastPage = false
                currentPage = START_PAGE

                repoAdapter.clear()
                fetchRepos(apiQuery!!, currentPage)

                return true
            }

            override fun onQueryTextChange(string: String): Boolean = true
        })

        searchView.setOnSearchClickListener {
            val query = queryPrefs.storedQuery
            searchView.setQuery(query, false)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean =
        when (item.itemId) {
            R.id.sort_by -> {
                val intent = Intent(activity, SettingsActivity::class.java)
                startActivity(intent)
                super.onOptionsItemSelected(item)
            }
            else -> super.onOptionsItemSelected(item)
        }

    override fun onAddFavoriteClick(repo: RepoEntry) {
        viewModel.addFavorite(repo)
        queryPrefs.lastResultDate = System.currentTimeMillis()
    }

    override fun onItemClick(repo: RepoEntry) {
        val intent = RepoReadmeActivity.newIntent(activity as Context, repo.owner.login, repo.name)
        startActivity(intent)
    }

    private fun fetchRepos(query: String, page: Int) {
        viewModel.fetchData(query, page)
    }

    private fun observeUI() {
        viewModel.loadingState.observe(viewLifecycleOwner, Observer { state ->
            isContentLoading = false

            when (state) {
                is RepoListViewModel.LoadingState.Loading -> {
                    commits_recycler_view?.visibility = View.INVISIBLE
                    load_indicator?.visibility = View.VISIBLE
                }
                is RepoListViewModel.LoadingState.Failed -> {
                    Toast.makeText(activity, "No repositories found for $apiQuery",
                            Toast.LENGTH_SHORT).show()
                    load_indicator?.visibility = View.GONE
                    Crashlytics.log(1, "RepoListFragment", "Retrofit error")
                    Crashlytics.logException(state.exception)
                }
                is RepoListViewModel.LoadingState.Succeeded -> {
                    commits_recycler_view?.visibility = View.VISIBLE
                    load_indicator?.visibility = View.GONE
                    processRepos(state.repos)
                }
            }
        })
    }

    private fun processRepos(repos: List<RepoEntry>?) {
        val apiQuery = queryPrefs.storedQuery ?: ""

        if (repos.isNullOrEmpty()) {
            Toast.makeText(activity, "No repositories found for $apiQuery",
                    Toast.LENGTH_SHORT).show()
            load_indicator?.visibility = View.GONE
            return
        } else {
            repoAdapter.apply {
                addAll(repos)
                sort()
            }
        }

        if (repos.size < MAXIMUM_LIST_LENGTH) {
            isContentLastPage = true
        }
    }

    companion object {
        private const val MAXIMUM_LIST_LENGTH = 30
        private const val START_PAGE = 1

        @JvmStatic
        fun newInstance(): RepoListFragment {
            return RepoListFragment()
        }
    }
}
